03/tideman/tideman.c

```diff
@@ -0,0 +1,232 @@
+        printf("Maximum number of candidates is %i\n", MAX);
+        return 2;
+    }
+    for (int i = 0; i < candidate_count; i++)
+    {
+        candidates[i] = argv[i + 1];
+    }
+
+    // Clear graph of locked in pairs
+    for (int i = 0; i < candidate_count; i++)
+    {
+        for (int j = 0; j < candidate_count; j++)
+        {
+            locked[i][j] = false;
+        }
+    }
+
+    pair_count = 0;
+    int voter_count = get_int("Number of voters: ");
+
+    // Query for votes
+    for (int i = 0; i < voter_count; i++)
+    {
+        // ranks[i] is voter's ith preference
+        int ranks[candidate_count];
+
+        // Query for each rank
+        for (int j = 0; j < candidate_count; j++)
+        {
+            string name = get_string("Rank %i: ", j + 1);
+
+            if (!vote(j, name, ranks))
+            {
+                printf("Invalid vote.\n");
+                return 3;
+            }
+        }
+
+        record_preferences(ranks);
+
+        printf("\n");
+    }
+
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+    return 0;
+}
+
+// Update ranks given a new vote
+bool vote(int rank, string name, int ranks[])
+{
+    for(int candidate = 0; candidate < candidate_count; candidate++) {
+
+        if(strcmp(name, candidates[candidate]) == 0) {
+            ranks[rank] = candidate;
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// Update preferences given one voter's ranks
+void record_preferences(int ranks[])
+{
+    for(int candidate_a = 0; candidate_a < candidate_count; candidate_a++) {
+
+        for(int candidate_b = candidate_a + 1; candidate_b < candidate_count; candidate_b++) {
+            preferences[ranks[candidate_a]][ranks[candidate_b]]++;
+        }
+    }
+
+    return;
+}
+
+// Record pairs of candidates where one is preferred over the other
+void add_pairs(void)
+{
+    for(int candidate_a = 0; candidate_a < candidate_count; candidate_a++) {
+
+        for(int candidate_b = candidate_a + 1; candidate_b < candidate_count; candidate_b++) {
+
+            if(preferences[candidate_a][candidate_b] > preferences[candidate_b][candidate_a]) {
+
+                pairs[pair_count].winner = candidate_a;
+                pairs[pair_count].loser = candidate_b;
+                pair_count++;
+            }
+            else if(preferences[candidate_a][candidate_b] < preferences[candidate_b][candidate_a]) {
+
+                pairs[pair_count].winner = candidate_b;
+                pairs[pair_count].loser = candidate_a;
+                pair_count++;
+            }
+        }
+    }
+
+    return;
+}
+
+// Sort pairs in decreasing order by strength of victory
+void sort_pairs(void)
+{
+    for(int outerCounter = 0; outerCounter < pair_count; outerCounter++) {
+
+        for(int innerCounter = outerCounter; innerCounter < pair_count; innerCounter++) {
+
+            int strengthCurrentPair = preferences[pairs[innerCounter].winner][pairs[innerCounter].loser];
```

Comment:
```suggestion
    for(int outerCounter = 0; outerCounter < pair_count; outerCounter++) {
        for(int innerCounter = outerCounter; innerCounter < pair_count; innerCounter++) {
            int strengthCurrentPair = preferences[pairs[innerCounter].winner][pairs[innerCounter].loser];
```
Direkt beim Beginn der Schleifen brauchst du keine leere Zeile